java/src/main/java/com/microsoft/onnxruntime/ONNXUtil.java

```diff
@@ -0,0 +1,648 @@
+        long elementCount = elementCount(shape);
+        // This check implicitly checks if it's a valid Java array, as otherwise it won't be comparable to input.length.
+        if (elementCount != input.length) {
+            throw new IllegalArgumentException("Expected " + elementCount + " elements to fill this shape, received " + input.length);
+        }
+
+        // This could be further optimised with use of System.arraycopy.
+        switch (shape.length) {
+            case 1: {
+                return Arrays.copyOf(input,input.length);
+            }
+            case 2: {
+                int dimOne = (int) shape[0];
+                int dimTwo = (int) shape[1];
+                int[][] output = new int[dimOne][dimTwo];
+
+                int count = 0;
+                for (int i = 0; i < dimOne; i++) {
+                    for (int j = 0; j < dimTwo; j++) {
+                        output[i][j] = input[count];
+                        count++;
+                    }
+                }
+
+                return output;
+            }
+            case 3: {
+                int dimOne = (int) shape[0];
+                int dimTwo = (int) shape[1];
+                int dimThree = (int) shape[2];
+                int[][][] output = new int[dimOne][dimTwo][dimThree];
+
+                int count = 0;
+                for (int i = 0; i < dimOne; i++) {
+                    for (int j = 0; j < dimTwo; j++) {
+                        for (int k = 0; k < dimThree; k++) {
+                            output[i][j][k] = input[count];
+                            count++;
+                        }
+                    }
+                }
+
+                return output;
+            }
+            case 4: {
+                int dimOne = (int) shape[0];
+                int dimTwo = (int) shape[1];
+                int dimThree = (int) shape[2];
+                int dimFour = (int) shape[3];
+                int[][][][] output = new int[dimOne][dimTwo][dimThree][dimFour];
+
+                int count = 0;
+                for (int i = 0; i < dimOne; i++) {
+                    for (int j = 0; j < dimTwo; j++) {
+                        for (int k = 0; k < dimThree; k++) {
+                            for (int l = 0; l < dimFour; l++) {
+                                output[i][j][k][l] = input[count];
+                                count++;
+                            }
+                        }
+                    }
+                }
+
+                return output;
+            }
+            case 5: {
+                int dimOne = (int) shape[0];
+                int dimTwo = (int) shape[1];
+                int dimThree = (int) shape[2];
+                int dimFour = (int) shape[3];
+                int dimFive = (int) shape[4];
+                int[][][][][] output = new int[dimOne][dimTwo][dimThree][dimFour][dimFive];
+
+                int count = 0;
+                for (int i = 0; i < dimOne; i++) {
```

Comment:
I am wondering, if data copying is actually necessary? Can you just replace the shape?

Reply:
No, unfortunately in Java a multidimensional array is made up of object arrays which point to object arrays which eventually point to primitive arrays. This is a terrible abstraction for numerics as a [10,200,200,3] batch of images is actually made up of 400,000 3 element arrays all allocated independently on the heap. Java needs a tensor type.

Reply:
It's not so hard to come up with an API that looks pretty enough and doesn't require copying everything. Take a look at JavaCPP's indexer package for an example:
http://bytedeco.org/news/2014/12/23/third-release/

Reply:
I feel strongly that we should implement it with nested or multidimensional arrays. We need a  continuous buffer. Would it be possible to expose an interface with the underlying original buffer?